### PR TITLE
[Fix #3570] avoid if-modifiers on multi-line statements

### LINKED ISSUE #3570: Guard clauses on multi-line statements
In doing a code review, I came across something like the following:

~~~ ruby
      foo[:bar] = {
        # Several more lines
      } unless baz.nil?
~~~

I was rather surprised that RuboCop didn't flag that. Guard clauses are great, but if the guard clause is so far away from the beginning of the statement, it gets too confusing. This seems like a cop worth adding. (And I suppose there should be a corresponding change to the Ruby style guide.)

### PR DESCRIPTION
**Pull request by tessi**
This implements a Cop to handle [multiline if modifiers](https://github.com/bbatsov/ruby-style-guide#no-multiline-if-modifiers).

This is my first commit to this project, so I wouldn't mind some extra care before merging ;)
Especially the `autocorrect` I've implemented might have some edges I have not foreseen.

### CODE REVIEW

File: config/enabled.yml [hunk 1]
```diff
@@ -450,6 +450,11 @@ Style/MultilineIfThen:
   StyleGuide: '#no-then'
   Enabled: true
 
+Style/MultilineIfModifier:
+  Description: 'Ensures if/unless modifiers are only applied to single lines.'
```
Comment (scottmatthewman) [inline config/enabled.yml]: Generally, these descriptions are written to read as intructions on how to write more correct code. I'd suggest something like `Only use if/unless modifiers on single line statements`.

File: lib/rubocop/cop/style/multiline_if_modifier.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+module RuboCop
+  module Cop
+    module Style
+      # Checks for uses of if/unless guard clauses on multiple line statements
+      #
+      # @example
+      #
+      #   # bad
+      #   {
+      #     result: 'this should not happend'
```
Comment (scottmatthewman) [inline lib/rubocop/cop/style/multiline_if_modifier.rb]: Typo here: I think `happend` should be `happen`

File: lib/rubocop/cop/style/multiline_if_modifier.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      #   } unless cond
+      #
+      #   # good
+      #   { result: 'ok' } if cond
+      class MultilineIfModifier < Cop
+        include IfNode
+        include StatementModifier
+        include AutocorrectAlignment
+
+        MSG = 'Favor a normal %s-expression over a modifier' \
```
Comment (scottmatthewman) [inline lib/rubocop/cop/style/multiline_if_modifier.rb]: Elsewhere in the code base, we refer to `if-statement` and `unless-statement`, so we should probably use that wording here.

File: lib/rubocop/cop/style/multiline_if_modifier.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+        def autocorrect(node)
+          lambda do |corrector|
+            corrector.replace(node.source_range, to_normal_if(node))
+          end
+        end
+
+        def to_normal_if(node)
+          cond, body = if_node_parts(node)
+          indended_body = indended_body(body)
```
Comment (scottmatthewman) [inline lib/rubocop/cop/style/multiline_if_modifier.rb]: `indended_body` should be spelled `indented_body`

File: lib/rubocop/cop/style/multiline_if_modifier.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+          "#{node.loc.keyword.source} #{cond.source}\n#{indended_body}\nend"
+        end
+
+        def configured_indentation_width
+          super || 2
+        end
+
+        def indended_body(body)
+          indendation = SPACE * configured_indentation_width
```
Comment (scottmatthewman) [inline lib/rubocop/cop/style/multiline_if_modifier.rb]: `indendation` should be spelled `indentation`

**Comment by tessi** [comment]
Ha! Of course my cop introduced some failures on the rubocop codebase itself. Ideal debugging cases for me. I'm investigating.

**Comment by tessi** [comment]
Thanks @scottmatthewman for the review! I fixed my typos based on your comments.

In addition to that, the cop got some improvements in detecting offenses (multi-line conditions are OK, whereas multiline if-bodies are not) and auto-correction (better handling of code indentation).

Some places in the rubocop codebase were targeted by the new cop. Those are fixed as well.

**Comment by bbatsov** [comment]
I just wonder if this should be a separate cop or part of the existing cop about modifier expressions. @jonas054 @alexdowad Any thoughts?

**Comment by alexdowad** [comment]
@bbatsov I don't see any other cop which looks like it overlaps with the intent of this one. `IfUnlessModifier` is quite different; perhaps it should be given a more descriptive name.

@tessi has some code in there for adjusting the indentation of autocorrected code... I feel like we should have a generic helper method for that somewhere already. Or if we don't, we should.

@tessi Nice work! One thing you still need to do is to squash these commits.

**Comment by Drenmi** [comment]
> @tessi has some code in there for adjusting the indentation of autocorrected code... I feel like we should have a generic helper method for that somewhere already. Or if we don't, we should.

I am inclined to say that this should be done by another cop (leveraging the inspection loop), or this concern will start to tangle itself into a lot of cops whose primary concerns should be something else. WDYT?

File: spec/rubocop/cop/style/multiline_if_modifier_spec.rb [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+      include_examples 'no offense'
+    end
+
+    context 'multiline condition' do
+      let(:source) { "run if cond &&\n       cond2" }
+
+      include_examples 'no offense'
+    end
+
+    context 'indeted offense' do
```
Comment (tessi) [inline spec/rubocop/cop/style/multiline_if_modifier_spec.rb]: argh, for some reason I fall easy to typos with that word

**Comment by tessi** [comment]
Thanks for your feedback. I'll rebase and squash my commits in a minute.

Regarding the indentation code, I had similar thoughts: First, I assumed some other Cop will take care of indentation -- which is actually the case. Unfortunately, this produced some additional whitespace warnings (when using the auto-correct functionality) -- but those whitespace error were not the users fault (but the fault of my Cop) -- and I felt bad for the user. On top of that, the produced indentation was a little off (the `end` was not indented properly). The latter could probably be fixed in some whitespace-Cop, the former issue might be harder to fix.
Anyhow, I thought that a cop's autocorrect should produce 'good' code from the start, which is why I decided to implement proper indentation in my Cop.

I assumed there was be a "indent my random piece of code automatically"-helper, but found none. So, I re-used some functions of the `AutocorrectAlignment` module, which was the closest thing I could find (apparently, implementing indentation has its own pitfalls -- for example that the user can configure the indentation width).

IMO cops should produce 'good' code when auto-correcting (including proper indentation), which is why we should have some helpers to indent code. The question is if those should be written in this PR (I'm willing to try) or not. I'd say it should be done in a later PR and be reviewed separately.

**Comment by bbatsov** [comment]
> IMO cops should produce 'good' code when auto-correcting (including proper indentation), which is why we should have some helpers to indent code. The question is if those should be written in this PR (I'm willing to try) or not. I'd say it should be done in a later PR and be reviewed separately.

Ideally, yes, but I wouldn't say that's a major issue. Keep in mind that sometimes the proper generated code would be dependent on RuboCop's config.

**Comment by tessi** [comment]
> Ideally, yes, but I wouldn't say that's a major issue. Keep in mind that sometimes the proper generated code would be dependent on RuboCop's config.

That's right. It's probably unfeasible to follow (and maintain) all configs in every cop.
Just to be clear: do you suggest to remove the indentation parts of this PR? If so, I'd remove it and rebase.
Are there any other things I should do before merging?

**Comment by Drenmi** [comment]
> IMO cops should produce 'good' code when auto-correcting (including proper indentation), which is why we should have some helpers to indent code.

I foresee the helpers would almost immediately become a shortcut to "run all the indentation cops on this code." 😀

**Comment by bbatsov** [comment]
> Just to be clear: do you suggest to remove the indentation parts of this PR? If so, I'd remove it and rebase.

I guess that's fine.

**Pull request merged**